File: src/ipc/curation.ts
// IPC alan modulu: kurasyon — etiket/favori/koleksiyon/proje-durumu (+ toplu kurasyon),
// asset iliskileri, RAG-dislama ve cop kutusu (soft-delete).
// Tuketiciler dogrudan buradan degil, `./client` facade'inden import eder.

import { invoke } from "@tauri-apps/api/core";
import type { AssetRow, CollectionRef, Facet, ProjectMeta } from "./assets";

/** TOPLU proje-durum yamasi — `bulk_set_project_meta` argumani (camelCase; sunucu `BulkProjectMetaArg`
 *  ile birebir). Her alan icin `apply*` bayragi: yalniz true olan YAZILIR (isaretlenmeyen korunur —
 *  tek-asset `ProjectMeta` full-replace'inden fark). Deger bos/whitespace → sunucuda null (temizlenir).
 *  `rejectionReason` yalniz onay uygulanip 'rejected' iken yazilir (aksi halde temizlenir). */
export interface BulkProjectPatch {
  applyClient: boolean;
  clientName: string | null;
  applyApproval: boolean;
  approvalStatus: string | null;
  rejectionReason: string | null;
  applyVersion: boolean;
  versionLabel: string | null;
  applyDeadline: boolean;
  deadline: string | null; // ISO YYYY-MM-DD
}

/** Bir onay durumu GECISI (asset-bazli gecmis; sunucu `ApprovalLogRow`, camelCase). En yeni once.
 *  `fromStatus`/`toStatus` null olabilir (durum atanmamis). `reason` yalniz reddedilmede dolu.
 *  `changedAt` unix SANIYE (JS Date icin ×1000). Detay panelindeki "Onay geçmişi" bunu gosterir. */
export interface ApprovalLogEntry {
  id: number;
  fromStatus: string | null;
  toStatus: string | null;
  reason: string | null;
  changedBy: string;
  changedAt: number;
}

/** Asset iliski turu (urun-kumesi; backend dogrular). `duplicate` kopya · `version` surum ·
 *  `xref` dis referans (XREF/link) · `derived` turetilmis (render/PDF ciktisi) · `backup` yedek/kilit
 *  sidecar (.bak/.dwl → kaynak). */
export type RelationKind = "duplicate" | "version" | "xref" | "derived" | "backup";

/** Bir asset iliskisi (detay "İlişkiler" sekmesi; sunucu `RelationDto`, camelCase). Yon iki
 *  taraflidir: `outgoing=true` → BU asset kaynak (→ karsi); `false` → karsi kaynak (← bu).
 *  `other*` alanlari bagli (karsi) AKTIF asset'i tanimlar (coptekiler backend'de haric). */
export interface Relation {
  id: number;
  kind: RelationKind;
  outgoing: boolean;
  otherId: number;
  otherPath: string;
  otherFileName: string;
}

/** Bir PROJE (entity) satiri — adli proje; asset'ler `assets.project_id` ile ona atanir.
 *  Sunucu `ProjectRow` ile birebir (camelCase). `assetCount`: bu projeye atanmis AKTIF
 *  (coptekiler haric) asset sayisi. Opsiyonel alanlar backend'de NULL → burada null. */
export interface ProjectRow {
  id: number;
  name: string;
  clientName: string | null;
  phase: string | null;
  status: string | null;
  description: string | null;
  deadline: string | null; // ISO YYYY-MM-DD
  createdAt: number; // unix saniye
  assetCount: number;
}

/** Proje olustur/guncelle girdisi (sunucu `ProjectInput` ile birebir; camelCase). `name`
 *  ZORUNLU (bos → backend Err). Opsiyonel alanlar bos ise **null** gonderilir ("" DEGIL) →
 *  backend NULL saklar. `update_project` full-replace'tir (verilmeyen alan null'lanir). */
export interface ProjectInput {
  name: string;
  clientName?: string | null;
  phase?: string | null;
  status?: string | null;
  description?: string | null;
  deadline?: string | null; // ISO YYYY-MM-DD
}

/** Aynı-kök OTO-tespit raporu (`detect_relations`; camelCase). */
export interface DetectRelationsReport {
  /** Aday grup sayısı (≥2 farklı uzantılı aynı-kök). */
  groups: number;
  /** Yeni oluşturulan `derived` iliski sayısı. */
  created: number;
}

/** Kurasyon (etiket/favori/koleksiyon/proje-durum/iliski/cop) komut sarmalayicilari — facade `ipc`'ye yayilir. */
export const curationIpc = {
  // ── Toplu kurasyon (P2.5): undo-kayitli, tek komut (per-asset dongu yerine) ──
  /** Favori durumunu TOPLU ayarla (delta; yalniz degisen id kaydedilir → undo hassas). Editor+.
   *  Doner: degisen sayi. */
  bulkSetFavorite: (ids: number[], on: boolean): Promise<number> =>
    invoke<number>("bulk_set_favorite", { ids, on }),

  /** Etiketi TOPLU ekle (delta; undo-kayitli). Editor+. Doner: degisen sayi. */
  bulkAddTag: (ids: number[], name: string): Promise<number> =>
    invoke<number>("bulk_add_tag", { ids, name }),

  /** Etiketi TOPLU kaldir (delta; undo-kayitli). Editor+. Doner: degisen sayi. */
  bulkRemoveTag: (ids: number[], name: string): Promise<number> =>
    invoke<number>("bulk_remove_tag", { ids, name }),

  /** Ada gore (find-or-create) koleksiyona TOPLU ekle (undo-kayitli; koleksiyon bu islemle
   *  olustuysa geri-al onu bos kalinca siler). Editor+. Doner: koleksiyon id. */
  bulkAddToCollection: (ids: number[], name: string): Promise<number> =>
    invoke<number>("bulk_add_to_collection", { ids, name }),

  /** Koleksiyondan TOPLU cikar (undo-kayitli). Editor+. `name` yalniz undo panel etiketi. */
  bulkRemoveFromCollection: (collectionId: number, ids: number[], name: string): Promise<number> =>
    invoke<number>("bulk_remove_from_collection", { collectionId, ids, name }),

  // ── Kurasyon (Faz 4.3): etiket + favori (write = editor/admin) ──
  addTag: (assetId: number, name: string): Promise<void> =>
    invoke<void>("add_tag", { assetId, name }),

  removeTag: (assetId: number, name: string): Promise<void> =>
    invoke<void>("remove_tag", { assetId, name }),

  setTagColor: (name: string, color: string | null): Promise<void> =>
    invoke<void>("set_tag_color", { name, color }),

  /** Etiketi VARLIK duzeyinde yeniden adlandir (tum asset'lerde birden). Editor+.
   *  Hedef ad kullanimdaysa backend REDDEDER (sessiz birlestirme yok). Geri-alinabilir. */
  renameTag: (oldName: string, newName: string): Promise<void> =>
    invoke<void>("rename_tag", { oldName, newName }),

  /** Etiketi VARLIK duzeyinde sil → doner: etkilenen dosya sayisi. Editor+.
   *  Yikici ama geri-alinabilir (renk + tum baglar undo kaydinda anlik-goruntulenir). */
  deleteTag: (name: string): Promise<number> => invoke<number>("delete_tag", { name }),

  setFavorite: (assetId: number, on: boolean): Promise<void> =>
    invoke<void>("set_favorite", { assetId, on }),

  tagFacets: (limit: number): Promise<Facet[]> =>
    invoke<Facet[]>("tag_facets", { limit }),

  favoriteCount: (): Promise<number> => invoke<number>("favorite_count"),

  // ── Proje-durum (H2 pariti): musteri/onay/versiyon/teslim (write = editor/admin) ──
  /** Bir asset'in proje-durum alanlarini ayarla. Editor+ gerekir. Bos/whitespace
   *  string sunucuda null'a temizlenir; `approval_status` whitelist disi → hata.
   *  `ProjectMeta` (snake_case okuma DTO'su) → sunucu `ProjectMetaArg` (camelCase)
   *  alan adlarina cevrilir; boylece okuma/yazma tek tipi paylasir. */
  setProjectMeta: (assetId: number, meta: ProjectMeta): Promise<void> =>
    invoke<void>("set_project_meta", {
      assetId,
      meta: {
        clientName: meta.client_name,
        approvalStatus: meta.approval_status,
        rejectionReason: meta.rejection_reason,
        versionLabel: meta.version_label,
        deadline: meta.deadline,
      },
    }),

  /** Bir asset KUMESININ proje-durum alanlarini TOPLU ayarla. Editor+ gerekir. Yalniz
   *  `apply*=true` alanlar yazilir (isaretlenmeyen korunur); bos/whitespace sunucuda null'a
   *  temizlenir; `approvalStatus` whitelist disi → hata. `patch` zaten camelCase (sunucu
   *  `BulkProjectMetaArg` ile birebir). Doner: guncellenen satir sayisi. */
  bulkSetProjectMeta: (ids: number[], patch: BulkProjectPatch): Promise<number> =>
    invoke<number>("bulk_set_project_meta", { ids, patch }),

  /** Bir asset'in ONAY GECIS gecmisi (en yeni once; en cok 100). Salt-okuma (her rol).
   *  Onay durumu her degistiginde bir satir (eski→yeni + sebep + kim + ne zaman). */
  listApprovalLog: (assetId: number): Promise<ApprovalLogEntry[]> =>
    invoke<ApprovalLogEntry[]>("list_approval_log", { assetId }),

  /** Onay durumuna gore asset sayilari (proje-durum faceti). `extFacets`/`tagFacets`
   *  ile ayni `Facet` sekli. Salt-okuma (her rol). */
  approvalFacets: (): Promise<Facet[]> => invoke<Facet[]>("approval_facets"),

  /** Musteri adina gore asset sayilari (proje-durum faceti). Salt-okuma (her rol). */
  clientFacets: (limit: number): Promise<Facet[]> =>
    invoke<Facet[]>("client_facets", { limit }),

  /** Versiyon etiketine gore asset sayilari (proje-durum faceti). Salt-okuma (her rol). */
  versionFacets: (limit: number): Promise<Facet[]> =>
    invoke<Facet[]>("version_facets", { limit }),

  /** Termin yilina gore asset sayilari (proje-durum faceti). Salt-okuma (her rol). */
  deadlineYearFacets: (): Promise<Facet[]> => invoke<Facet[]>("deadline_year_facets"),

  // ── Koleksiyonlar (Faz 4.3b): adli asset gruplari (write = editor/admin) ──
  listCollections: (): Promise<CollectionRef[]> =>
    invoke<CollectionRef[]>("list_collections"),

  createCollection: (name: string): Promise<number> =>
    invoke<number>("create_collection", { name }),

  deleteCollection: (id: number): Promise<void> =>
    invoke<void>("delete_collection", { id }),

  renameCollection: (id: number, name: string): Promise<void> =>
    invoke<void>("rename_collection", { id, name }),

  setCollectionColor: (id: number, color: string | null): Promise<void> =>
    invoke<void>("set_collection_color", { id, color }),

  addToCollection: (collectionId: number, assetId: number): Promise<void> =>
    invoke<void>("add_to_collection", { collectionId, assetId }),

  removeFromCollection: (collectionId: number, assetId: number): Promise<void> =>
    invoke<void>("remove_from_collection", { collectionId, assetId }),

  // ── Projeler (entity): adli proje + asset↔proje atama (write = editor+; list = her rol) ──
  /** Yeni proje olustur (find-or-create DEGIL — her cagri yeni kayit). **Editor+** gerekir.
   *  Bos ad → backend Err. Doner: yeni proje id. */
  createProject: (input: ProjectInput): Promise<number> =>
    invoke<number>("create_project", { input }),

  /** Tum projeler + atanmis AKTIF asset sayilari (`assetCount`). Salt-okuma (her rol). */
  listProjects: (): Promise<ProjectRow[]> => invoke<ProjectRow[]>("list_projects"),

  /** Projeyi guncelle (ad + meta FULL-REPLACE; verilmeyen opsiyonel alan null'lanir).
   *  **Editor+** gerekir. Bos ad → backend Err. */
  updateProject: (id: number, input: ProjectInput): Promise<void> =>
    invoke<void>("update_project", { id, input }),

  /** Projeyi sil. **Editor+** gerekir. FK SET NULL: atanmis asset'ler SILINMEZ — yalniz
   *  proje atamalari kalkar (project_id → NULL). */
  deleteProject: (id: number): Promise<void> => invoke<void>("delete_project", { id }),

  /** Asset KUMESINI bir projeye ata (`assets.project_id`). **Editor+** gerekir.
   *  `projectId=null` → atamayi KALDIR. Olmayan projeye atama → backend Err.
   *  Doner: etkilenen asset sayisi. */
  assignAssetsToProject: (ids: number[], projectId: number | null): Promise<number> =>
    invoke<number>("assign_assets_to_project", { ids, projectId }),

  // ── Cop kutusu / soft-delete (§O): sil→cop, listele, geri-yukle, kalici-sil ──
  /** Asset'leri cop kutusuna at (soft-delete; GERI-ALINABILIR). Editor+ gerekir.
   *  Etkilenen (cop'e atilan) sayisi doner. */
  trashAssets: (ids: number[]): Promise<number> =>
    invoke<number>("trash_assets", { ids }),

  /** Asset'leri cop kutusundan geri yukle (restore; yeniden aktif). Editor+ gerekir.
   *  Etkilenen (geri donen) sayisi doner. */
  restoreAssets: (ids: number[]): Promise<number> =>
    invoke<number>("restore_assets", { ids }),

  /** Asset'leri KALICI sil (purge; GERI-ALINAMAZ — tum iliskili veri silinir).
   *  **Admin** gerekir. Etkilenen sayisi doner. */
  purgeAssets: (ids: number[]): Promise<number> =>
    invoke<number>("purge_assets", { ids }),

  /** Cop kutusundaki asset'ler (en son atilan ilk). `list_assets` ile ayni `AssetRow`
   *  sekli. Salt-okuma (her rol). */
  listTrash: (): Promise<AssetRow[]> => invoke<AssetRow[]>("list_trash"),

  /** Cop kutusundaki asset sayisi (TopBar rozeti). Salt-okuma (her rol). */
  trashCount: (): Promise<number> => invoke<number>("trash_count"),

  // ── §G Asset iliskileri (detay "İlişkiler" sekmesi): oku her rol; ekle/kaldir editor+ ──
  /** Bir asset'in iliskileri (her iki yon; karsi AKTIF asset'e cozulmus). Salt-okuma (her rol).
   *  Copteki karsi-uclar backend'de haric. */
  assetRelations: (assetId: number): Promise<Relation[]> =>
    invoke<Relation[]>("asset_relations", { assetId }),

  /** İki asset arasi iliski kur (`srcId` → `dstId`, `kind`). **Editor+** gerekir. `kind` urun-kumesi
   *  disi → backend hata; `src==dst` reddedilir; tekrar → no-op (backend UNIQUE). */
  addRelation: (srcId: number, dstId: number, kind: RelationKind): Promise<void> =>
    invoke<void>("add_relation", { srcId, dstId, kind }),

  /** Bir iliskiyi id ile kaldir. **Editor+** gerekir. */
  removeRelation: (id: number): Promise<void> =>
    invoke<void>("remove_relation", { id }),

  /** Aynı-kök (plan.dwg ↔ plan.pdf) iliskilerini OTO-tespit et (`detect_relations`; **admin**).
   *  Farkli-uzantili ayni-kok gruplarini `derived` bagler; mevcut/manuel baglari EZMEZ. Rapor doner. */
  detectRelations: (): Promise<DetectRelationsReport> =>
    invoke<DetectRelationsReport>("detect_relations"),

  /** Asset'leri RAG'den manuel disla / dahil et (A1 hassasiyet manuel istisnasi). Editor+ gerekir.
   *  `excluded=true` → sohbet retrieve'i artik bu asset'leri getirmez (kalici; re-index korur).
   *  Etkilenen asset sayisi doner. */
  setRagExcluded: (ids: number[], excluded: boolean): Promise<number> =>
    invoke<number>("set_rag_excluded", { ids, excluded }),
};

File: src-tauri/src/commands/curation.rs
//! Kullanici kurasyonu: etiket + favori (Faz 4.3) ve koleksiyonlar (Faz 4.3b).
//! Yazma = editor+; sayim/liste salt-okuma (her rol).

use crate::rbac;
use crate::AppState;
use archivist_db::CollectionRef;
use tauri::State;

// ── Kullanici kurasyonu (Faz 4.3): etiket + favori (editor+ gate) ────────────

/// Bir asset'e kullanici etiketi ekle. **Editor/Admin** (yazma).
#[tauri::command(async)]
pub fn add_tag(
    asset_id: i64,
    name: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let role = rbac::current_role(&state).map_err(|e| e.to_string())?;
    rbac::require_editor(role).map_err(|e| e.to_string())?;
    let mut db = state.db.lock().map_err(|e| e.to_string())?;
    db.add_user_tag(asset_id, &name).map_err(|e| e.to_string())
}

/// Bir asset'ten etiket bagini kaldir. **Editor/Admin** (yazma).
#[tauri::command(async)]
pub fn remove_tag(
    asset_id: i64,
    name: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let role = rbac::current_role(&state).map_err(|e| e.to_string())?;
    rbac::require_editor(role).map_err(|e| e.to_string())?;
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.remove_tag(asset_id, &name).map_err(|e| e.to_string())
}

/// Kullanici etiketinin global rengini ayarla veya temizle. **Editor/Admin**.
#[tauri::command(async)]
pub fn set_tag_color(
    name: String,
    color: Option<String>,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let role = rbac::current_role(&state).map_err(|e| e.to_string())?;
    rbac::require_editor(role).map_err(|e| e.to_string())?;
    let actor = crate::audit::actor(&state);
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.set_tag_color(&name, color.as_deref()).map_err(|e| e.to_string())?;
    crate::audit::record_on(&db, &actor, "tag_recolor", Some("tag"), Some(name.trim()), color.as_deref());
    Ok(())
}

/// Etiketi **yeniden adlandir** (varlik duzeyinde; tum asset'lerde birden). **Editor/Admin**.
///
/// H2 `commandRenameTag` pariteli (2026-07-26 davranis-sadakati turu §8). Hedef ad zaten
/// kullanimdaysa DB katmani reddeder (sessiz birlestirme YOK). Geri-alinabilir.
#[tauri::command(async)]
pub fn rename_tag(
    old_name: String,
    new_name: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let role = rbac::current_role(&state).map_err(|e| e.to_string())?;
    rbac::require_editor(role).map_err(|e| e.to_string())?;
    let actor = crate::audit::actor(&state);
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.rename_user_tag(&old_name, &new_name).map_err(|e| e.to_string())?;
    crate::undo_commands::record_tag_rename(&db, old_name.trim(), new_name.trim());
    crate::audit::record_on(
        &db,
        &actor,
        "tag_rename",
        Some("tag"),
        Some(old_name.trim()),
        Some(new_name.trim()),
    );
    Ok(())
}

/// Etiketi **sil** (varlik duzeyinde; tum asset'lerden kalkar) → doner: etkilenen dosya sayisi.
/// **Editor/Admin**. H2 `commandDeleteTag` pariteli.
///
/// Yikici ama **geri-alinabilir**: silmeden ONCE renk + bagli asset id'leri anlik-goruntulenir ve
/// `undo_ops`'a yazilir → "Geri Al" etiketi rengiyle ve tum baglariyla geri kurar (H2 `restoreTag`).
/// UI ayrica onay diyalogu gosterir (kac dosyadan kalkacagi yazili).
#[tauri::command(async)]
pub fn delete_tag(name: String, state: State<'_, AppState>) -> Result<i64, String> {
    let role = rbac::current_role(&state).map_err(|e| e.to_string())?;
    rbac::require_editor(role).map_err(|e| e.to_string())?;
    let actor = crate::audit::actor(&state);
    let db = state.db.lock().map_err(|e| e.to_string())?;
    // Anlik-goruntu SILMEDEN ONCE (sonra okunamaz).
    let color = db.tag_color(&name).map_err(|e| e.to_string())?;
    let ids = db.delete_user_tag(&name).map_err(|e| e.to_string())?;
    crate::undo_commands::record_tag_delete(&db, name.trim(), color.as_deref(), &ids);
    crate::audit::record_on(
        &db,
        &actor,
        "tag_delete",
        Some("tag"),
        Some(name.trim()),
        Some(&ids.len().to_string()),
    );
    Ok(ids.len() as i64)
}

/// Favori durumunu ayarla (true=ekle, false=kaldir). **Editor/Admin** (yazma).
#[tauri::command(async)]
pub fn set_favorite(
    asset_id: i64,
    on: bool,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let role = rbac::current_role(&state).map_err(|e| e.to_string())?;
    rbac::require_editor(role).map_err(|e| e.to_string())?;
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.set_favorite(asset_id, on).map_err(|e| e.to_string())
}

/// Favori asset sayisi (filtre rozeti). Salt-okuma (her rol).
#[tauri::command(async)]
pub fn favorite_count(state: State<'_, AppState>) -> Result<i64, String> {
    let db = state.read_db.lock().map_err(|e| e.to_string())?;
    db.favorite_count().map_err(|e| e.to_string())
}

// ── Koleksiyonlar (Faz 4.3b): adli asset gruplari (yazma = editor+) ──────────

/// Tum koleksiyonlar + uye sayilari. Salt-okuma (her rol).
#[tauri::command(async)]
pub fn list_collections(state: State<'_, AppState>) -> Result<Vec<CollectionRef>, String> {
    let db = state.read_db.lock().map_err(|e| e.to_string())?;
    db.list_collections().map_err(|e| e.to_string())
}

/// Koleksiyon olustur (find-or-create; id doner). **Editor/Admin** (yazma).
#[tauri::command(async)]
pub fn create_collection(name: String, state: State<'_, AppState>) -> Result<i64, String> {
    let role = rbac::current_role(&state).map_err(|e| e.to_string())?;
    rbac::require_editor(role).map_err(|e| e.to_string())?;
    let mut db = state.db.lock().map_err(|e| e.to_string())?;
    db.create_collection(&name, None).map_err(|e| e.to_string())
}

/// Koleksiyonu sil (uyelikler CASCADE; asset'lere dokunmaz). **Editor/Admin** (yazma).
#[tauri::command(async)]
pub fn delete_collection(id: i64, state: State<'_, AppState>) -> Result<(), String> {
    let role = rbac::current_role(&state).map_err(|e| e.to_string())?;
    rbac::require_editor(role).map_err(|e| e.to_string())?;
    let actor = crate::audit::actor(&state);
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.delete_collection(id).map_err(|e| e.to_string())?;
    crate::audit::record_on(&db, &actor, "collection_delete", Some("collection"), Some(&id.to_string()), None);
    Ok(())
}

/// Koleksiyonu yeniden adlandirir. Uyelikler koleksiyon kimligiyle bagli oldugu icin korunur.
/// **Editor/Admin** gerekir; audit kaydina yeni ad yazilir.
#[tauri::command(async)]
pub fn rename_collection(id: i64, name: String, state: State<'_, AppState>) -> Result<(), String> {
    let role = rbac::current_role(&state).map_err(|e| e.to_string())?;
    rbac::require_editor(role).map_err(|e| e.to_string())?;
    let actor = crate::audit::actor(&state);
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.rename_collection(id, &name).map_err(|e| e.to_string())?;
    crate::audit::record_on(
        &db,
        &actor,
        "collection_rename",
        Some("collection"),
        Some(&id.to_string()),
        Some(name.trim()),
    );
    Ok(())
}

/// Koleksiyonun renk rozetini ayarlar veya kaldirir. **Editor/Admin** gerekir.
#[tauri::command(async)]
pub fn set_collection_color(
    id: i64,
    color: Option<String>,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let role = rbac::current_role(&state).map_err(|e| e.to_string())?;
    rbac::require_editor(role).map_err(|e| e.to_string())?;
    let actor = crate::audit::actor(&state);
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.set_collection_color(id, color.as_deref()).map_err(|e| e.to_string())?;
    crate::audit::record_on(
        &db,
        &actor,
        "collection_recolor",
        Some("collection"),
        Some(&id.to_string()),
        color.as_deref(),
    );
    Ok(())
}

/// Bir asset'i koleksiyona ekle. **Editor/Admin** (yazma).
#[tauri::command(async)]
pub fn add_to_collection(
    collection_id: i64,
    asset_id: i64,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let role = rbac::current_role(&state).map_err(|e| e.to_string())?;
    rbac::require_editor(role).map_err(|e| e.to_string())?;
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.add_to_collection(collection_id, asset_id).map_err(|e| e.to_string())
}

/// Bir asset'i koleksiyondan cikar. **Editor/Admin** (yazma).
#[tauri::command(async)]
pub fn remove_from_collection(
    collection_id: i64,
    asset_id: i64,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let role = rbac::current_role(&state).map_err(|e| e.to_string())?;
    rbac::require_editor(role).map_err(|e| e.to_string())?;
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.remove_from_collection(collection_id, asset_id).map_err(|e| e.to_string())
}
